contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+      acct.weights[wIndex].weight = approvalWeight;
+
+      const deltaApprovalWeight = api.BigNumber(approvalWeight)
+        .minus(oldApprovalWeight)
+        .dp(token.precision, api.BigNumber.ROUND_HALF_UP);
+
+      if (!api.BigNumber(deltaApprovalWeight).eq(0)) {
+        await api.db.update('accounts', acct);
+        const approvals = await api.db.find('approvals',
+          { from: account, candidatePending: true },
+          params.maxAccountApprovals,
+          0,
+          [{ index: '_id', descending: true }]);
+        for (let index = 0; index < approvals.length; index += 1) {
+          const approval = approvals[index];
+          const candidatePending = await updateCandidateWeight(approval.to,
+            deltaApprovalWeight,
+            token);
+          if (!candidatePending) {
+            approval.candidatePending = false;
+            await api.db.update('approvals', approval);
+          }
+        }
+      }
+    }
+  }
+};
+
+// ticks
+
+async function payRecipient(account, symbol, quantity, type = 'user', contractPayload = null) {
+  if (api.BigNumber(quantity).gt(0)) {
+    const res = await api.transferTokens(account, symbol, quantity, type);
+    if (type === 'contract' && contractPayload) {
+      await api.executeSmartContract(account, 'receiveRolesTokens',
+        { data: contractPayload, symbol, quantity });
+    }
+    if (res.errors) {
+      api.debug(`Error paying out roles of ${quantity} ${symbol} to ${account} (TXID ${api.transactionId}): \n${res.errors}`);
+      return false;
+    }
+    return true;
+  }
+  return false;
+}
+
+async function checkPendingCandidates(inst, params) {
+  const random = api.random();
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const upInst = JSON.parse(JSON.stringify(inst));
+  const voteTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: inst.voteToken });
+  const voteTokenMinValue = api.BigNumber(1)
+    .dividedBy(api.BigNumber(10).pow(voteTokenObj.precision));
+  const instTickTime = api.BigNumber(blockDate.getTime())
+    .minus(params.instanceTickHours * 3600 * 1000).toNumber();
+
+  let rolesProcessed = 0;
+  const pendingRoles = await api.db.find('roles',
+    {
+      instanceId: inst.id,
+      active: true,
+      'tokenBalances.0': { $exists: true },
+      lastTickTime: { $lte: instTickTime },
+    },
+    params.maxRolesPerBlock,
+    0,
+    [{ index: 'byLastTickTime', descending: false }, { index: '_id', descending: false }]);
+
+  for (let i = 0; i < pendingRoles.length; i += 1) {
+    const role = pendingRoles[i];
+    const funded = [];
+    const payTokens = role.tokenBalances.filter(t => api.BigNumber(t.quantity).gt(0));
+    const totalSlots = api.BigNumber(role.mainSlots).plus(role.backupSlots).toNumber();
+    const roleTickTime = api.BigNumber(blockDate.getTime())
+      .minus(role.tickHours * 3600 * 1000).toNumber();
+
+    if (role.lastTickTime <= roleTickTime) {
```

Comment:
this could be part of the query as well? currently it checks <= instTickTime but we can have it be <= min(two tick times)